lib/dom.js

```diff
@@ -4,6 +4,117 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
+/* Crypto */
+
+// This is a helper type to simplify the specification, it isn't an interface
+// and there are no objects implementing it.
+// https://developer.mozilla.org/en-US/docs/Web/API/BufferSource
+type BufferSource = ArrayBuffer | $ArrayBufferView;
+
+type DOMString = string;
+
+declare interface Crypto extends RandomSource {
+  subtle: SubtleCrypto; // readonly
+}
+
+type CryptoKey$Type =
+  | 'secret'
+  | 'public'
+  | 'private';
+type CryptoKey$Usages =
+  | 'encrypt'
+  | 'decrypt'
+  | 'sign'
+  | 'verify'
+  | 'deriveKey'
+  | 'deriveBits'
+  | 'wrapKey'
+  | 'unwrapKey';
+declare interface CryptoKey {
+    algorithm: Object; // readonly
+    extractable: Boolean; // readonly
+    type: CryptoKey$Type; // readonly
+    usages: CryptoKey$Usages[]; // readonly
+}
+
+// TODO: CryptoKeyPair
+declare interface CryptoKeyPair {
+
+}
+
+// TODO: JSONWebKey
+type JSONWebKey = {};
+
+type RandomSource =  {
+  getRandomValues(typedArray: $TypedArray): void;
+};
+
+type SubtleCrypto$Algorithm =
+  | {name: 'AES-CBC', iv: BufferSource}
+  | {name: 'AES-CTR', counter: BufferSource, length: number}
+  | {name: 'AES-GCM', iv: BufferSource, additionalData?: BufferSource, tagLength?: number}
+  | {name: 'RSA-OAEP', label?: string};
+type SubtleCrypto$DerivedKeyAlgorithm = {
+  name: DOMString,
+  length?: 128 | 192 | 256,
+};
+// TODO: fix types for parameters
+type SubtleCrypto$DeriveKeyAlgorithm =
+  | {name: 'ECDH', public: any}
+  | {name: 'DH', public: any}
+  | {name: 'PBKDF2', salt: BufferSource, iterations: number, hash: DOMString}
+  | {name: 'HKDF-CTR', hash: any, label: any, context: any};
+type SubtleCrypto$GenerateKeyAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW'
+  | 'HMAC'
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA'
+  | 'ECDH'
+  | 'DH';
+type SubtleCrypto$HashAlgo =
+  | 'SHA-1'
+  | 'SHA-256'
+  | 'SHA-384'
+  | 'SHA-512';
+type SubtleCrypto$KeyFormat =
+  | 'raw'
+  | 'pkcs8'
+  | 'spki'
+  | 'jwk';
+type SubtleCrypto$SignAlgo =
+  | 'HMAC'
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA';
+type SubtleCrypto$WrapAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW';
+declare interface SubtleCrypto {
+  decrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<string>;
+  // TODO: fix type for 'algorithm' param
+  deriveBits(algorithm: any, baseKey: CryptoKey, length: number): Promise<ArrayBuffer>;
+  deriveKey(algorithm: SubtleCrypto$DeriveKeyAlgorithm, masterKey: CryptoKey, derivedKeyAlgorithm: DOMString | SubtleCrypto$DerivedKeyAlgorithm, extractable: Boolean, keyUsages: CryptoKey$Usages[]): Promise<CryptoKey | CryptoKeyPair>;
+  digest(algo: SubtleCrypto$HashAlgo, buffer: BufferSource): Promise<string>;
```

Comment:
This definitely returns something `ArrayBuffer`-like and not a string.

Reply:
This is was not defined in any spec, but it should be a buffer indeed 🙂
`BufferSource` is more appropriate I think, since the input (`buffer`) is of type `BufferSource`?

Reply:
Regarding all methods:
Saying it is a `BufferSource` won't break anything, but I don't see why any vendor would return an array representation of a buffer with a particular bitness and not a naked buffer.
I suspect `ArrayBuffer` is more correct for the returns.
`BufferSource` is indeed the type for all inputs, even `decrypt`.

Reply:
OK will make them `ArrayBuffer` for now 🙂
We can always change them to `BufferSource` if necessary 🙂